VultisigApp/VultisigApp/Services/1inch/OneInchService.swift

```diff
@@ -36,26 +36,57 @@ struct OneInchService {
             fee: 0.5,
             isAffiliate: isAffiliate
         )
-
+        
         var request = URLRequest(url: url)
         request.allHTTPHeaderFields = [
             "accept": "application/json",
         ]
-
+        
         let (data, _) = try await URLSession.shared.data(for: request)
         let response = try JSONDecoder().decode(OneInchQuote.self, from: data)
         return response
     }
-
+    
     func fetchTokens(chain: Int) async throws -> [OneInchToken] {
         let response: OneInchTokensResponse = try await Utils.fetchObject(from: Endpoint.fetchTokens(chain: chain))
         let tokens = Array(arrayLiteral: response.tokens.values).reduce([], +)
         return tokens
     }
+    
+    func fetchTokensBalance(chainId: Int, walletAddress: String) async throws -> [OneInchToken] {
+        let oneInchChainId = String(describing: chainId)
+        let balanceData = try await Utils.asyncGetRequest(urlString: Endpoint.fetch1InchsTokensBalance(chain: oneInchChainId, address: walletAddress), headers: [:])
+        
+        if let balanceDictionary = try JSONSerialization.jsonObject(with: balanceData, options: []) as? [String: String] {
+            
+            let tokensWithNonZeroBalance = balanceDictionary.filter { tokenAddress, balance in
+                if let balanceValue = Int64(balance) {
+                    return balanceValue > 0
+                }
+                return false
+            }
+            
+            let nonZeroBalanceTokenAddresses = tokensWithNonZeroBalance.map { $0.key }
+            
+            let tokenInfoData = try await Utils.asyncGetRequest(
+                urlString: Endpoint.fetch1InchsTokensInfo(chain: oneInchChainId, addresses: nonZeroBalanceTokenAddresses), headers: [:]
+            )
+            
+            var tokens: [OneInchToken] = []
+            for tokenAddress in nonZeroBalanceTokenAddresses {
+                if let tokenInfo: OneInchToken = Utils.extractResultFromJson(fromData: tokenInfoData, path: tokenAddress, type: OneInchToken.self) {
+                    tokens.append(tokenInfo)
+                }
+            }
+            return tokens;
+        }
+        return []
```

Comment:
This is throwing function. Can we do not return empty array if something went wrong in parsing?

Reply:
Yep. We are allowing it to throw, we only return [] if the user has no tokens or if any of them has no balance, therefore if any error happens on try ... it is throwing.